source/localizable/tutorial/service.md

```diff
@@ -0,0 +1,293 @@
+{{else}}
+  <button {{action "imageShow"}}>Show image</button>
+{{/if}}
+{{location-map location=rental.city}}
+```
+
+Next, generate the map component using Ember-CLI.
+
+```shell
+ember g component location-map
+```
+
+Running this command generates three files: a component JavaScript file, a template, and a test file.
+To help think through what we want our component to do, we'll implement a test case first.
+
+In this case we plan on having our Google Maps service handle map display.  
+Our component's job will be to take the results from the map service (which is a map element) and append it to an element in the component template.
+
+All we want to do in our component is make sure its appending the output of the maps service to an inner map container element.
+To limit the test to validating just this behavior we'll take advantage of the registration API to provide a mock maps service.
+In the mock service, define a method that will fetch the map based on location, called `getMapElement`.
+
+```tests/integration/components/location-map-test.js
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import Ember from 'ember';
+
+let MockMapsService = Ember.Service.extend({
+  getMapElement(location) {
+    this.set('calledWithLocation', location);
+    // We create a div here to simulate our maps service,
+    // which will create and then cache the map element
+    return document.createElement('div');
+  }
+});
+
+moduleForComponent('location-map', 'Integration | Component | location map', {
+  integration: true,
+  beforeEach() {
+    this.register('service:maps', MockMapsService);
+    this.inject.service('maps', { as: 'mapsService' });
+  }
+});
+
+test('should append map element to container element', function(assert) {
+  this.set('myLocation', 'New York');
+  this.render(hbs`{{location-map location=myLocation}}`);
+  assert.equal(this.$('.map-container').children().length, 1, 'container should have one child');
+  assert.equal(this.get('mapsService.calledWithLocation'), 'New York', 'should call service with New York');
+});
+```
+
+Notice that our mock service above sets a variable called `calledWithLocation`.  We later assert on this variable to ensure that the service is being called with the same location provided to the component.
+
+To get the test to pass, add the container element to the component template.
+
+```app/templates/components/location-map.hbs
+<div class="map-container"></div>
+```
+
+Then update the component to append the map output to our component.
+We'll add a maps service injection, and call the `getMapElement` function with the provided location.
+
+```app/components/location-map.js
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+  maps: Ember.inject.service(),
+
+  didInsertElement() {
+    this._super(...arguments);
+    let location = this.get('location');
+    let mapElement = this.get('maps').getMapElement(location)
```

Comment:
Thanks for providing input, but I'm not sure what you are asking.  Generally with services, you use the inject helper to make the service available for an object, but that service isn't available until it initially gets called with get.  I agree in this case that the component is getting the service pretty soon after inject, but this is just showing a simple case.  If you really wanted to go further and show off lazy loading, you could take the map lookup out of the didInsertElement and base it off of the component showing up in the viewport, but that would be beyond what we want to show in the tutorial.